test/test_distributed.py

```diff
@@ -1366,13 +1367,69 @@ def test_all_gather_cuda(self):
     @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
     def test_all_gather_group(self):
         group, group_id, rank = self._init_group_test()
+        print("group", group, group_id)
         self._test_all_gather_helper(group, group_id, rank)
 
     @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
     def test_all_gather_full_group(self):
         group, group_id, rank = self._init_full_group_test()
         self._test_all_gather_helper(group, group_id, rank)
 
+
+    def _test_all_gather_coalesced_helper(
+        self, group, group_id, rank
+    ):
+        # TODO: Instead we should probably go through _rank_not_in_group
+        # mechanism to disable sending tensors
+        if group_id is not None:
+            for test_case_id in range(2, 5):
+                # Make sure we create tensors of incompatible sizes, e.g.
+                # [1x1], [2x2], ... to be sent in one batch
+                input_tensors = [
+                    _build_tensor(
+                        tensor_id, rank + tensor_id) for tensor_id in range(
+                            1, test_case_id)
+                ]
+                output_tensor_lists = [
+                    [
+                        _build_tensor(tensor_id, -1) for tensor_id in range(
+                            1, test_case_id)
+                    ] for _ in group
+                ]
+                dist.all_gather_coalesced(
+                    output_tensor_lists, input_tensors, group_id)
+
+                expected_tensors = [
+                    [
+                        _build_tensor(
+                            tensor_id,
+                            rank_iter + tensor_id) for tensor_id in range(
+                                1, test_case_id)
+                    ] for rank_iter in group
+                ]
+                for l1, l2 in zip(output_tensor_lists, expected_tensors):
+                    for t1, t2 in zip(l1, l2):
+                        assert torch.equal(t1, t2)
+
+        self._barrier()
+
+    @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
+    def test_all_gather_coalesced(self):
+        group, group_id, rank = self._init_global_test()
+        self._test_all_gather_coalesced_helper(group, group_id, rank)
+
+    @skip_if_small_worldsize
+    @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
+    def test_all_gather_coalesced_group(self):
+        group, group_id, rank = self._init_group_test()
+        print(group, group_id)
+        self._test_all_gather_coalesced_helper(group, group_id, rank)
+
+    @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
+    def test_all_gather_coalesced_full_group(self):
+        group, group_id, rank = self._init_full_group_test()
+        self._test_all_gather_coalesced_helper(group, group_id, rank)
+
```

Comment:
Let's add tests to make sure that when sizes input and output tensor lists do not match, we throw error with clear error message.

Reply:
Done